### PR TITLE
RESTWS-916: Create an endpoint to expose the NameTemplate resource as a REST document

### PR DESCRIPTION
**Pull request by IamMujuziMoses**
## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-916

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java [hunk 1]
```diff
@@ -273,9 +274,8 @@ protected PageableResult doSearch(RequestContext context) throws ResponseExcepti
 			else {
 				return new NeedsPaging<Order>(orders, context);
 			}
```
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java:276]: Is this change related to the pull request?

File: omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/OrderController1_10Test.java [hunk 2]
```diff
@@ -589,4 +590,12 @@ public void invalidCareCenterShouldThrowException() throws Exception {
 		        );
 		handle(req);
 	}
+
```
Comment (dkayiwa) [inline omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/OrderController1_10Test.java:593]: Is this change related to the pull request?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Override
+    public void globalPropertyDeleted(String s) {
+        if (getNameFormat().equals(s))
+            this.setNameTemplate((List)(new Vector()));
+    }
+
+    public String getNameFormat() {
+
+        return "<org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0>" +
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:121]: Where did you get this from?

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 1 above)
Comment (IamMujuziMoses) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java:276]: No, it got mixed up from another branch, let me see how I can fix that

File: omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/OrderController1_10Test.java (hunk 2 above)
Comment (IamMujuziMoses) [reply omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/OrderController1_10Test.java:593]: No, it got mixed up from another branch, let me see how I can fix that

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java (hunk 3 above)
Comment (IamMujuziMoses) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:121]: That is the String for the "layout.name.format" value which didn't exist

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            String[] tokens = customDefaults.split(",");
+            Map<String, String> elementDefaults = nameTemplate.getElementDefaults();
+
+            for (String token : tokens) {
+                String[] pair = token.split("=");
+                if (pair.length == 2) {
+                    String name = pair[0];
+                    String val = pair[1];
+
+                    if (elementDefaults == null) {
```
Comment (tendomart) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java:58]: you wanna do something similar here

`if ( StringUtils.isEmpty(elementDefaults)) {`

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0;
+import org.openmrs.serialization.SerializationException;
+
+public class NameSupport2_0 extends LayoutSupport<NameTemplate2_0> implements GlobalPropertyListener {
+    private static NameSupport2_0 singleton;
+    private boolean initialized = false;
+    static Log log = LogFactory.getLog(NameSupport2_0.class);
+
+    private NameSupport2_0() {
+        if (singleton == null)
```
Comment (tendomart) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:32]: `if (StringUtils.isEmpty(singleton))`

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/template/openmrs2_0/NameTemplate2_0.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        return "IS_NOT_NAME_TOKEN";
+    }
+
+    @Override
+    public LayoutSupport<?> getLayoutSupportInstance() {
+        return NameSupport2_0.getInstance();
+    }
+
+    public String getCodeName() {
+        if (this.codeName == null) {
```
Comment (tendomart) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/template/openmrs2_0/NameTemplate2_0.java:41]: `if (StringUtils.isEmpty(this.codename)) {`

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 1 above)
Comment (IamMujuziMoses) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java:276]: @dkayiwa is there any way I can drop the first commit (RESTWS-716) from this pull request?!

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            String[] tokens = customDefaults.split(",");
+            Map<String, String> elementDefaults = nameTemplate.getElementDefaults();
+
+            for (String token : tokens) {
+                String[] pair = token.split("=");
+                if (pair.length == 2) {
+                    String name = pair[0];
+                    String val = pair[1];
+
+                    if (elementDefaults.isEmpty()) {
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java:58]: Is there a possibility of this being NULL?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NameSupport2_0 extends LayoutSupport<NameTemplate2_0> implements GlobalPropertyListener {
+    private static NameSupport2_0 singleton;
+    private boolean initialized = false;
+    static Logger log = LoggerFactory.getLogger(NameSupport2_0.class);
+
+    private NameSupport2_0() {
+        if (Objects.isNull(singleton))
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:34]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            singleton = this;
+        log.debug("Setting singleton: {}", singleton);
+    }
+
+    public static NameSupport2_0 getInstance() {
+        synchronized (NameSupport2_0.class) {
+            if (Objects.isNull(singleton))
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:41]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        return this.defaultLayoutFormat;
+    }
+
+    public void setNameTemplate(List<NameTemplate2_0> nameTemplates) {
+        this.layoutTemplates = nameTemplates;
+        this.setDefaultLayoutFormat(this.layoutTemplates.get(0).getCodeName());
+    }
+
+    public List<NameTemplate2_0> getNameTemplate() {
+        if (CollectionUtils.isEmpty(this.layoutTemplates))
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:74]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        List<NameTemplate2_0> list = new ArrayList<>();
+        list.add(nameTemplate);
+        this.setNameTemplate(list);
+    }
+
+    @Override
+    public void globalPropertyDeleted(String s) {
+        if (getNameFormat().equals(s))
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:117]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java (hunk 8 above)
Comment (IamMujuziMoses) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:34]: okay, let me make the changes accordingly

**Comment by dkayiwa** [comment]
@IamMujuziMoses i was asked for a status update of this by @vasharma05

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0;
+import org.openmrs.serialization.SerializationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NameSupport2_0 extends LayoutSupport<NameTemplate2_0> implements GlobalPropertyListener {
+    private static NameSupport2_0 singleton;
+    private boolean initialized = false;
+    static Logger log = LoggerFactory.getLogger(NameSupport2_0.class);
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java:31]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-HowToLog

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/nametemplate")
+public class NameTemplateController2_0  extends BaseRestController {
+
+    private static final Logger log = LoggerFactory.getLogger(NameTemplateController2_0.class);
+
+    @RequestMapping(method = RequestMethod.GET)
+    @ResponseBody
+    public Object get(WebRequest request) throws SerializationException {
+
+        NameTemplate2_0 nameTemplate = NameSupport2_0.getInstance().getDefaultLayoutTemplate();
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0.java:44]: Is there any reason why you did not use `NameSupport.getInstance().getDefaultLayoutTemplate()`?

File: omod-2.0/src/test/resources/nameTemplate.xml [hunk 14]
```diff
@@ -0,0 +1,15 @@
+<org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0>
```
Comment (dkayiwa) [inline omod-2.0/src/test/resources/nameTemplate.xml:1]: Why are we using `NameTemplate2_0` instead of `NameTemplate`?

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/RestConstants.java [hunk 15]
```diff
@@ -201,7 +201,9 @@ public class RestConstants {
 	public static boolean SWAGGER_LOGS_ON = true;
 	
 	public static boolean SWAGGER_LOGS_OFF = false;
-	
+
+	public static final String LAYOUT_NAME_FORMAT = "<org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0><nameMappings class=\"properties\"><property name=\"familyName\" value=\"Person.familyName\"/><property name=\"middleName\" value=\"Person.middleName\"/><property name=\"givenName\" value=\"Person.givenName\"/></nameMappings><sizeMappings class=\"properties\"><property name=\"familyName\" value=\"40\"/><property name=\"middleName\" value=\"40\"/><property name=\"givenName\" value=\"40\"/></sizeMappings><lineByLineFormat><string>familyName middleName givenName</string></lineByLineFormat></org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0>";
```
Comment (dkayiwa) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/RestConstants.java:205]: What are we using this for?

File: omod-2.0/src/test/resources/nameTemplate.xml (hunk 14 above)
Comment (IamMujuziMoses) [reply omod-2.0/src/test/resources/nameTemplate.xml]: I had not seen the `NameTemplate` but I rectified this in the last commit

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/template/openmrs2_0/NameTemplate2_0.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+package org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0;
+
+import java.io.Serializable;
+
+import org.apache.commons.lang3.StringUtils;
+import org.openmrs.layout.LayoutSupport;
+import org.openmrs.layout.LayoutTemplate;
+import org.openmrs.module.webservices.rest.web.v1_0.support.openmrs2_0.NameSupport2_0;
+
+public class NameTemplate2_0 extends LayoutTemplate implements Serializable {
```
[the remark this answers is not in the document]
Comment (IamMujuziMoses) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/template/openmrs2_0/NameTemplate2_0.java]: I had not seen the `NameTemplate` class so I thought of creating a new one, but this is rectified in the last commit

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+import org.openmrs.api.GlobalPropertyListener;
+import org.openmrs.api.context.Context;
+import org.openmrs.layout.LayoutSupport;
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.v1_0.template.openmrs2_0.NameTemplate2_0;
+import org.openmrs.serialization.SerializationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NameSupport2_0 extends LayoutSupport<NameTemplate2_0> implements GlobalPropertyListener {
```
[the remark this answers is not in the document]
Comment (IamMujuziMoses) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/support/openmrs2_0/NameSupport2_0.java]: I had not seen the `NameSupport` class so I thought of creating a new one, but this is rectified in the last commit
